pkg/middlewares/aws_defaults.go

```diff
@@ -0,0 +1,181 @@
+package middlewares
+
+import (
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/cloudskiff/driftctl/pkg/resource"
+	"github.com/cloudskiff/driftctl/pkg/resource/aws"
+)
+
+const defaultIamRolePathPrefix = "/aws-service-role/"
+
+// When scanning a brand new AWS account, some users may see irrelevant results about default AWS role policies.
+// We ignore these resources by default when strict mode is disabled.
+type AwsDefaults struct{}
+
+func NewAwsDefaults() AwsDefaults {
+	return AwsDefaults{}
+}
+
+func awsIamRoleDefaults(remoteResources, resourcesFromState *[]resource.Resource) error {
+	newRemoteResources := make([]resource.Resource, 0)
+
+	for _, remoteResource := range *remoteResources {
+		// Ignore all resources other than iam role
+		if remoteResource.TerraformType() != aws.AwsIamRoleResourceType {
+			newRemoteResources = append(newRemoteResources, remoteResource)
+			continue
+		}
+
+		existInState := false
+		for _, stateResource := range *resourcesFromState {
+			if resource.IsSameResource(remoteResource, stateResource) {
+				existInState = true
+				break
+			}
+		}
+
+		if existInState {
+			newRemoteResources = append(newRemoteResources, remoteResource)
+			continue
+		}
+
+		if match := strings.HasPrefix(*remoteResource.(*aws.AwsIamRole).Path, defaultIamRolePathPrefix); !match {
+			newRemoteResources = append(newRemoteResources, remoteResource)
+			continue
+		}
+
+		logrus.WithFields(logrus.Fields{
+			"id":   remoteResource.TerraformId(),
+			"type": remoteResource.TerraformType(),
+		}).Debug("Ignoring IAM role as it is a resource managed by the cloud provider")
+	}
+
+	*remoteResources = newRemoteResources
+
+	return nil
+}
+
+func awsIamPolicyAttachmentDefaults(remoteResources, resourcesFromState *[]resource.Resource) error {
+	newRemoteResources := make([]resource.Resource, 0)
+
+	for _, remoteResource := range *remoteResources {
+		// Ignore all resources other than iam policy attachment
+		if remoteResource.TerraformType() != aws.AwsIamPolicyAttachmentResourceType {
+			newRemoteResources = append(newRemoteResources, remoteResource)
+			continue
+		}
+
+		existInState := false
+		for _, stateResource := range *resourcesFromState {
+			if resource.IsSameResource(remoteResource, stateResource) {
+				existInState = true
+				break
+			}
+		}
+
+		if existInState {
+			newRemoteResources = append(newRemoteResources, remoteResource)
+			continue
+		}
+
+		for _, roleId := range *remoteResource.(*aws.AwsIamPolicyAttachment).Roles {
+			var role *aws.AwsIamRole
+			for _, res := range *remoteResources {
+				if res.TerraformType() == aws.AwsIamRoleResourceType && res.TerraformId() == roleId {
+					role = res.(*aws.AwsIamRole)
+				}
+			}
+
+			// If we couldn't find the linked role, include the resource
```

Comment:
I'm not sure this is the expected behavior, in theory it should never happen as an attachment cannot exist without a role.

WDYT @wbeuil I'm not clear about this

Reply:
Since we're looping across remote resources, indeed, it should never happen. Good catch.